### PR TITLE
Ensure outbound tracing headers take precedence

### LINKED ISSUE #682: InjectOutboundSpan doesn't Override Tracing Header
Not sure if it is expected behavior or a bug, but [`InjectOutboundSpan`](https://github.com/uber/tchannel-go/blob/dev/tracing.go#L172) would not inject if the input parameter `headers` already contains tracing information using the same tracing header.
[tracing.go#L187](https://github.com/uber/tchannel-go/blob/dev/tracing.go#L187)

Currently to inject tracing header correctly to outbound headers, we need to do the following:
```
delete(headers, "XX-Tracing-Header")
headers := tchannel.InjectOutboundSpan(outboundCall.Response(), headers)
```

### PR DESCRIPTION
**Pull request by yurishkuro**
Resolves #682 

Signed-off-by: Yuri Shkuro <ys@uber.com>

cc @Yu-Xie

### CODE REVIEW

**Comment by yurishkuro** [comment]
keep forgetting that `dev` is the trunk branch

File: tracing.go [hunk 1]
```diff
@@ -185,7 +185,11 @@ func InjectOutboundSpan(response *OutboundCallResponse, headers map[string]strin
 		return headers // Tracer did not add any tracing headers, so return the original map
 	}
 	for k, v := range headers {
-		newHeaders[k] = v
+		// Some applications propagate all inbound application headers to outbound calls (issue #682).
+		// If those headers include tracing headers we want to make sure to keep the new tracing headers.
+		if _, ok := newHeaders[k]; !ok {
+			newHeaders[k] = v
```
Comment (prashantv) [inline tracing.go]: Do you know why we didn't get coverage from the added test here?

Comment (yurishkuro) [reply tracing.go]: Don't know. When I run this locally it shows the line as covered:
```
$ go test -v -run TestTracingSpanAttributes -coverprofile cover.out
=== RUN   TestTracingSpanAttributes
--- PASS: TestTracingSpanAttributes (0.00s)
PASS
coverage: 48.8% of statements
ok  	github.com/uber/tchannel-go	0.042s

$ go tool cover -html=cover.out
```

image

I added another check to ensure that non-tracing headers are propagated.

File: tracing_test.go [hunk 2]
```diff
@@ -60,23 +65,47 @@ func TestTracingSpanAttributes(t *testing.T) {
 		DisableRelay:   true,
 	}
 	WithVerifiedServer(t, opts, func(ch *Channel, hostPort string) {
+		const (
+			customAppHeaderKey           = "futurama"
+			customAppHeaderExpectedValue = "simpsons"
+		)
+		var customAppHeaderValue atomic.Value
```
Comment (prashantv) [inline tracing_test.go]: we use `uber-go/atomic` in tchannel, which has a `atomic.String` that you could use here

File: tracing_test.go [hunk 3]
```diff
@@ -60,23 +65,47 @@ func TestTracingSpanAttributes(t *testing.T) {
...
 		span := ch.Tracer().StartSpan("client")
 		span.SetBaggageItem(testtracing.BaggageKey, testtracing.BaggageValue)
 		ctx := opentracing.ContextWithSpan(context.Background(), span)
 		root := new(testtracing.TracingResponse).ObserveSpan(ctx)
 
+		// Pretend that the client propagated tracing headers from upstream call, and test that the outbound call
+		// will override them (https://github.com/uber/tchannel-go/issues/682).
+		headers := make(map[string]string)
+		assert.NoError(t, tracer.Inject(span.Context(), opentracing.TextMap, opentracing.TextMapCarrier(headers)))
+		tracingHeaders := map[string]string{
```
Comment (prashantv) [inline tracing_test.go]: should these be named the opposite? this seems like `headers`, while the thing named `headers` seems more like `tracingHeaders`

File: tracing_test.go [hunk 4]
```diff
@@ -60,23 +65,47 @@ func TestTracingSpanAttributes(t *testing.T) {
...
+			customAppHeaderKey: customAppHeaderExpectedValue,
+		}
+		for k, v := range headers {
+			tracingHeaders["$tracing$"+k] = v
```
Comment (prashantv) [inline tracing_test.go]: can you set the value to something incorrect, like `dummy` so we know that it was replaced

**Comment by Yu-Xie** [comment]
Thanks Yuri for fixing this! Do we have a plan to release?

**Comment by prashantv** [comment]
@Yu-Xie This is now released in the latest release:
https://github.com/uber/tchannel-go/releases/tag/v1.10.0

**Comment by Yu-Xie** [comment]
Thanks @prashantv

**Pull request merged**